Deprecate accounts

jnewbery opened the pull request:
Deprecate all accounts functionality and make it only accessible by using `-deprecatedrpc=accounts`.

Accounts specific RPCs, account arguments, and account related results all require the `-deprecatedrpc=accunts` startup option now in order to see account things.

Several wallet functional tests use the accounts system. Those tests are unchanged, except to start the nodes with `-deprecatedrpc=accounts`. We can slowly migrate those tests to use the 'label' API instead of the 'account' API before accounts are fully removed.

jnewbery committed:
[tests] Rename rpc_listtransactions.py to wallet_listtransactions.py

listtransactions is a wallet RPC. The test name should indicate that
this is a wallet test.

jnewbery commented:
This is @achow101's #11497 rebased and reworked. Notable differences:

- I've moved the RPC method deprecation up into the `CRPCTable::execute()` function. That's perhaps a little bit messy (rpc/server.cpp ideally wouldn't have knowledge of individual Bitcoin RPCs), but it saves a lot of code duplication and will be removed in V0.18, so I decided it was acceptable to have it there.
- Some methods (`sendmany` and `listtransactions`) have alternative help text depending on whether the `-deprecatedrpcs=accounts` switch is used. That means some duplicate text, but it makes the help text a lot clearer in all cases (since it's not filed with "if using accounts this else that").
- I've avoided the argument repositioning that achow101 used, since that breaks the way rpc server fills in positional arguments from named arguments. That means that `sendmany` and `listtransactions` now have a dummy argument in the same way that `move` did. We can do another API change in a future version if that's offensive to people.

This PR also renames `rpc_listtransactions.py` to `wallet_listtransactions.py` (since it's a wallet test). This makes it clear that only wallet tests are affected by this change.

kallewoof commented:
Review hint: recommend append `?w=1` to end, to avoid having to review whitespace changes.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1039,13 +1044,54 @@ UniValue sendmany(const JSONRPCRequest& request)
-    if (request.fHelp || request.params.size() < 2 || request.params.size() > 8)
-        throw std::runtime_error(
-            "sendmany \"fromaccount\" {\"address\":amount,...} ( minconf \"comment\" [\"address\",...] replaceable conf_target \"estimate_mode\")\n"
+    std::string help_text {};
```
Why ` {}`?

jnewbery replied:
not required. I've now removed.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1039,13 +1044,54 @@ UniValue sendmany(const JSONRPCRequest& request)
             + HelpRequiringPassphrase(pwallet) + "\n"
             "\nArguments:\n"
-            "1. \"fromaccount\"         (string, required) DEPRECATED. The account to send the funds from. Should be \"\" for the default account\n"
+            "1. \"fromaccount\"         (string, required) DEPRECATED. The account to send the funds from. Should be \"\" for the default account.\n"
```
Remove the added `.` at end?

jnewbery replied:
removed

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1769,19 +1820,64 @@ UniValue listtransactions(const JSONRPCRequest& request)
+            "Note that the \"account\" argument and \"otheraccount\" return value have been removed in V0.17. To use this RPC with an \"account\" argument, restart\n"
+            "bitcoind with -deprecatedrpc=accounts\n"
+            "\nArguments:\n"
+            "1. \"dummy\"    (string, optional) If set, should be \"*\" for backwards compatibility.\n"
```
Since we're splitting the function anyway, why have the dummy at all?

I also wonder if we should have some form of upgrade plan for when we do this. I know e.g. @luke-jr prefers option objects.

promag replied:
I agree, I don't see a reason to support the case "I want the new behavior without accounts but keep the RPC argument". People that wants to avoid breaking change launch with `-deprecatedrpc=accounts`.

jnewbery replied:
@promag - please see https://github.com/bitcoin/bitcoin/pull/12953#issuecomment-380813038 for rationale.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1836,8 +1933,15 @@ UniValue listtransactions(const JSONRPCRequest& request)
-    if (!request.params[0].isNull())
-        strAccount = request.params[0].get_str();
+    if (!IsDeprecatedRPCEnabled("accounts")) {
+        if (request.params[0].get_str() != "*") {
```
If `params[0]` is unset, this will throw an error at `get_str()` won't it?
Looks like you want
```C++
if (!request.params[0].isNull()) {
  strAccount = request.params[0].get_str();
  if (!IsDeprecatedRPCEnabled("accounts") && strAccount != "*") {
    throw ...
  }
}
```

promag replied:
Correct, `accounts` is optional.

jnewbery replied:
Yes, much better. Thanks

kallewoof reviewed (commented):
utACK

laanwj commented:
(Obvious) concept ACK

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1039,13 +1044,54 @@ UniValue sendmany(const JSONRPCRequest& request)
+            "\nSend two amounts to two different addresses setting the confirmation and comment:\n"
+            + HelpExampleCli("sendmany", "\"{\\\"1D1ZrZNe3JUo7ZycKEYQQiQAWd9y54F4XX\\\":0.01,\\\"1353tsE8YMTA4EuV7dgUXGjNFf9KpVvKHz\\\":0.02}\" 6 \"testing\"") +
+            "\nSend two amounts to two different addresses, subtract fee from amount:\n"
+            + HelpExampleCli("sendmany", "\"{\\\"1D1ZrZNe3JUo7ZycKEYQQiQAWd9y54F4XX\\\":0.01,\\\"1353tsE8YMTA4EuV7dgUXGjNFf9KpVvKHz\\\":0.02}\" 1 \"\" \"[\\\"1D1ZrZNe3JUo7ZycKEYQQiQAWd9y54F4XX\\\",\\\"1353tsE8YMTA4EuV7dgUXGjNFf9KpVvKHz\\\"]\"") +
```
Missing dummy `""`?

jnewbery replied:
thanks. Fixed

promag commented on src/rpc/server.cpp:
```diff
@@ -474,13 +474,23 @@ static inline JSONRPCRequest transformNamedArguments(const JSONRPCRequest& in, c
 
 UniValue CRPCTable::execute(const JSONRPCRequest &request) const
 {
+    static const std::vector<std::string> account_methods = {"getaccountaddress", "getaccount",
```
Alternative:
```cpp
static const std::set<std::string> accounts_methods { "getaccountaddress", "getaccount", ... };
...

if (!IsDeprecatedRPCEnabled("accounts") && account_methods.count(request.strMethod)) {
```

jnewbery replied:
yes, that's nicer. Thanks

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1039,13 +1044,54 @@ UniValue sendmany(const JSONRPCRequest& request)
+            "\nSend two amounts to two different addresses, subtract fee from amount:\n"
+            + HelpExampleCli("sendmany", "\"{\\\"1D1ZrZNe3JUo7ZycKEYQQiQAWd9y54F4XX\\\":0.01,\\\"1353tsE8YMTA4EuV7dgUXGjNFf9KpVvKHz\\\":0.02}\" 1 \"\" \"[\\\"1D1ZrZNe3JUo7ZycKEYQQiQAWd9y54F4XX\\\",\\\"1353tsE8YMTA4EuV7dgUXGjNFf9KpVvKHz\\\"]\"") +
+            "\nAs a json rpc call\n"
+            + HelpExampleRpc("sendmany", "{\"1D1ZrZNe3JUo7ZycKEYQQiQAWd9y54F4XX\":0.01,\"1353tsE8YMTA4EuV7dgUXGjNFf9KpVvKHz\":0.02}, 6, \"testing\"");
```
Same as above.

jnewbery replied:
fixed

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1093,6 +1141,9 @@ UniValue sendmany(const JSONRPCRequest& request)
         throw JSONRPCError(RPC_CLIENT_P2P_DISABLED, "Error: Peer-to-peer functionality missing or disabled");
     }
 
+    if (!IsDeprecatedRPCEnabled("accounts") && LabelFromValue(request.params[0]) != "") {
```
Remove `LabelFromValue` because it raises `JSONRPCError(RPC_WALLET_INVALID_LABEL_NAME, "Invalid label name")`. I suggest to just `... && !request.params[0].get_str().empty()`.

jnewbery replied:
agree. Thanks

promag reviewed (commented):
Partially tested, some comments.

jnewbery commented:
> Since we're splitting the function anyway, why have the dummy at all?

I mentioned this briefly in the PR text:

> I've avoided the argument repositioning that achow101 used, since that breaks the way rpc server fills in positional arguments from named arguments.

(@achow101's argument repositioning code is here: https://github.com/bitcoin/bitcoin/pull/11497/files#diff-df7d84ff2f53fcb2a0dc15a3a51e55ceR1086 for example)

In more detail:

- the rpc/server.cpp function `transformNamedArguments()` uses the list of arguments from `CRPCTable ` to convert a keyed map of arguments into a univalue array.
- That `CRPCTable` is constructed from the various `CRPCCommand` arrays in the rpc .cpp files. For `sendmany`, the entry is:
```
    { "wallet",             "sendmany",                         &sendmany,                      {"fromaccount|dummy","amounts","minconf","comment","subtractfeefrom","replaceable","conf_target","estimate_mode"} },
```
- that means that when called with named arguments, the rpc server will eg place the `fromaccount` named argument into index 0 in the array, `amount` into index 1, and so on.
- there's no way to change `CRPCTable` at runtime based on whether or not `-deprecatedrpc` was used. Changing that in rpc/server.cpp would be a very large change, and would involve placing more bitcoin-specific knowledge into the rpc server code.

Functionally, I think this is good behaviour anyway - users who were calling `sendmany` with an empty string as the `account` argument will not see any difference when they upgrade. Users who were using `sendmany` with accounts will see an immediate obvious break, and can either change to stop using accounts or use `-deprecatedrpc=accounts` for one more release.

jnewbery commented:
Thanks for the review @kallewoof and @promag! Your comments are addressed in the latest push.

This requires an update to the release notes. I'll do that once the new code/functionality is stable.

promag commented:
Thanks @jnewbery for the details.

IIUC:
 - accounts are deprecated but API wise the arguments are still there;
 - in 0.18 accounts will be gone even dummy arguments.

If that's the case then I'm sorry to NACK - it's not possible to refactor existing software to face 0.18. In 0.17 we should provide both scenarios API wise!

I haven't thought about the positional and named arguments thing with `CRPCTable` but I believe we should find a solution.

jnewbery commented:
> in 0.18 accounts will be gone even dummy arguments.

No - there's no firm plan to remove the dummy arguments. At some point, we *could* remove them, but that would be a breaking API change.

The API change in this PR should be minimally invasive. For users calling `sendmany` and `listtransactions` without accounts, there's no change in the way they need to call the RPC. Changing the argument ordering as you've suggested would require *all* users of those PRC methods to change their client code, regardless of whether they're currently using accounts.

Does that address your concern? If not, could you elaborate a bit more?

laanwj commented on src/rpc/server.cpp:
```diff
@@ -474,13 +474,22 @@ static inline JSONRPCRequest transformNamedArguments(const JSONRPCRequest& in, c
             throw JSONRPCError(RPC_IN_WARMUP, rpcWarmupStatus);
     }
 
+    // Accounts RPC methods are deprecated
```
This is a smaller change, however because `rpc/server.cpp` is supposed to not be specific to bitcoin, I'd prefer to put the deprecation check in the individual methods.

jnewbery replied:
I agree that it's more correct to have these in the individual methods, but see my rationale above:

> That's perhaps a little bit messy (rpc/server.cpp ideally wouldn't have knowledge of individual Bitcoin RPCs), but it saves a lot of code duplication and will be removed in V0.18, so I decided it was acceptable to have it there.

Are you strongly opposed to have this code in rpc/server.cpp, even if just temporarily? If so, I can change the PR to move this code to individual methods.

promag replied:
> but it saves a lot of code duplication

Well it's just 8 deprecated methods. I'm fine either way.

laanwj replied:
To be honest I prefer the duplication. If it's the about of work just say so and I'll provide a patch.

jnewbery replied:
>  If it's the about the amount of work

no, mostly amount of review. I'm happy to write the code if you're happy to review it :)

jnewbery committed:
[tests] Set -deprecatedrpc=accounts in tests

Future commits will deprecate the accounts RPC methods, arguments and
return objects. Set the -deprecatedrpc=accounts switch now so tests
don't break in intermediate commits.

jnewbery committed:
[wallet] [rpc] Deprecate account RPC methods

All account RPC methods are now deprecated and can only be called if
bitcoind has been started with the -deprecatedrpc=accounts switch.

Affected RPC methods are:

- getaccount
- getaccountaddress
- getaddressesbyaccount
- getreceivedbyaccount
- listaccouts
- listreceivedbyaccount
- move
- setaccount

jnewbery committed:
[wallet] [rpc] Deprecate wallet 'account' API

This commit finalizes the deprecation of the wallet 'accounts' API by
removing all account arguments and return values.

RPC behaviour is slightly different if the 'accounts' or 'labels' API is
being used. Those behaviour changes are fully documented in the RPC help
text.

jnewbery committed:
[wallet] [rpc] Remove duplicate entries in rpcwallet.cpp's CRPCCommand table

Remove duplicate listreceivedby{account,label} methods.

jnewbery commented:
Moved the RPC method deprecation into rpcwallet.cpp as requested by @laanwj . Also added tests for those deprecated methods.

Still to do:

- deprecation tests for the changes in _Deprecate wallet 'account' API_
- release notes

jnewbery committed:
[wallet] [tests] Add tests for accounts/labels APIs

jnewbery committed:
[docs] Add release notes for deprecated 'account' API

jnewbery commented:
Yes, `--coverage` caught a genuine bug. Hat tip to @jamesob .

Now fixed, and regression tests and release notes added.

laanwj commented:
utACK cead28bbecf032b49c01ca3ae4f47aa2536cbc55

red0bear commented:
i need know then .... in 0.18 there is no more -deprecatedrpc=accounts ?

red0bear commented:
well, i was looking to. just to clato clarify; as I could send from one label to another from 0.18 being that the sendfrom is no longer among the commands to accomplish this function. From 0.18 / 0.19 will bitcoin treat your wallet as only one person?

sipa commented:
@red0bear It *always* treated your wallet as one person. The only difference between `sendtoaddress` and `sendfrom` is that the latter also subtracted the amount being sent from the specified accounts' balance (but that was just virtual bean counter, it never had any relation to what UTXOs existed on-chain).

red0bear commented:
So in these versions i can't anymore sendfrom "fromaccount" "toaddress" amount will not be replaced by another command then. And only sendtoaddress limit label to only description wallet then.

kallewoof commented:
@red0bear Accounts were broken. Horribly. You are risking your funds every time you use them. Do not. This is why they were removed.

luke-jr commented:
Eh, they were never broken AFAIK. We just decided to stop supporting them.

red0bear commented:
@kallewoof , any external connection is a risk today.

markblundeberg commented:
( note, introduces race fixed in #13130 )

laanwj commented:
a race *in the python tests*, not in the binary, to be clear

inzider commented:
Im trying to start version 0.19.1 with  -deprecated=accounts so I can use sendfrom method.
Still, bitcoind-cli return method not found.

Those are useful options for many reason - making a "blind" remove is odd. 

What is the solution here!? Downgrade or im doing something wrong?

luke-jr commented:
@inzider Please open a question on StackExchange explaining what you want to do. Accounts are no longer supported.

jnewbery commented:
@inzider - you need to use the `-deprecatedrpc=accounts` command-line or config option.
